### PR TITLE
Fix incorrect conversion to long double.

### PR DESCRIPTION
**Pull request by racardoso**
This commit fixes issue #60.

Signed-off-by: Rogerio Alves <rcardoso@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by racardoso** [comment]
> This patch is removing function `__dpd_trunctdtf`

It's not supposed to remove `__dpd_trunctdtf` it only change the function to not extend removing the line `#include "extendddtf.c"`

**Comment by tuliom** [comment]
> It's not supposed to remove __dpd_trunctdtf it only change the function to not extend removing the line #include "extendddtf.c"

Could you elaborate, please?
The definition of `__dpd_trunctdtf` is in `extendddtf.c` and depends on macros `SRC`, `DEST` and `NAME` defined in `base-math/trunctdtf.c`.

After removing this include directive, where is `__dpd_trunctdtf` coming from?

**Comment by racardoso** [comment]
@tuliom In fact it doesn't pass ABI tests. I made some changes. I am just discarding the second part of the number for now. There's still some wrong results for long double but, at least it doesn't give NAN instead of INF on the results.

**Comment by racardoso** [comment]
@tuliom force pushed a new commit this will fix all long double issues on cast to overflow tests

File: base-math/extendddtf.c [hunk 1]
```diff
@@ -30,20 +30,63 @@
...
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -0.0 : 0.0;
+          }
+       }
+
 	df_part1 = a;			/* TD -> DF  */
 	dd_part1 = df_part1;		/* DF -> DD./TD.  */
-	dd_part2 = a - dd_part1;	/* DD./TD.  */
-	df_part2 = dd_part2;		/* DD/TD -> DF.  */
```
Comment (gftg85) [inline base-math/extendddtf.c]: The removal of these lines causes new warning messages:

```
In file included from /home/gabriel/src/libdfp/base-math/extendddtf.c:36:0:
/home/gabriel/src/libdfp/base-math/extendddtf.c: In function ‘__dpd_extendddtf’:
/home/gabriel/src/libdfp/base-math/extendddtf.c:43:21: warning: unused variable ‘dd_part2’ [-Wunused-variable]
  SRC_TYPE dd_part1, dd_part2;
                     ^
/home/gabriel/src/libdfp/include/convert.h:348:4: note: in definition of macro ‘CONVERT_WRAPPER’
    __VA_ARGS__  \
    ^~~~~~~~~~~
/home/gabriel/src/libdfp/base-math/extendddtf.c:43:11: warning: variable ‘dd_part1’ set but not used [-Wunused-but-set-variable]
  SRC_TYPE dd_part1, dd_part2;
           ^
/home/gabriel/src/libdfp/include/convert.h:348:4: note: in definition of macro ‘CONVERT_WRAPPER’
    __VA_ARGS__  \
    ^~~~~~~~~~~
/home/gabriel/src/libdfp/base-math/extendddtf.c:41:28: warning: unused variable ‘df_part2’ [-Wunused-variable]
  volatile double df_part1, df_part2;
                            ^
/home/gabriel/src/libdfp/include/convert.h:348:4: note: in definition of macro ‘CONVERT_WRAPPER’
    __VA_ARGS__  \
    ^~~~~~~~~~~
```

Comment (racardoso) [reply base-math/extendddtf.c]: @gftg85  When I rebased and force pushed I forgot to add those lines back. Remove those lines was the part of my first commit and should be not on this commit. I've made another push. I guess everything is fine now. No warning and no failing tests on bfp-conversions.

**Comment by tuliom** [comment]
> I'm getting a regression on `test-bfp-conversions`:
> 
> ...
> 
> Have you seen this, too?

I've seen this too and these results are really weird.
Firstly, take a look at the complete output:

```
3.402819e+38 = (long double) 3.402819e+38; /* _Decimal128 */ in: /home/tuliom/src/libdfp/tests/test-bfp-conversions.c: 601
70  Error:   Expected: "1.983701e+2613"
              Result:   "4.398000e-6158"
in: /home/tuliom/src/libdfp/tests/test-bfp-conversions.c:333
```

We can see the first line is printing identical values for both the long double and the _Decimal128 (minus errors that did not appear after string conversions).
But the expected and result outputs are completely different from what we expected.
Those lines are printed by macro `_VC_P` which is expanded in line 610:

```
      _VC_P(__FILE__,tdtfp->line, tdtfp->e, retval, tdtfp->format);
```
Both `tdtfp->e` and `retval` are long double, with `tdtfp->e = 3.402819e+38L`.
However `tdtfp->format = "%DDe"` and that's the format being used in `_VC_P` to print them.

ISO C11 states the following for `fprintf()`:
> If any argument is not the correct type for the corresponding conversion specification, the behavior is undefined.

In other words: test-bfp-conversions is broken and needs to be fixed and I'm working on a patch.

**Comment by racardoso** [comment]
> > I'm getting a regression on `test-bfp-conversions`:
> > ...
> > Have you seen this, too?
> 
> I've seen this too and these results are really weird.
> Firstly, take a look at the complete output:
> 
> ```
> 3.402819e+38 = (long double) 3.402819e+38; /* _Decimal128 */ in: /home/tuliom/src/libdfp/tests/test-bfp-conversions.c: 601
> 70  Error:   Expected: "1.983701e+2613"
>               Result:   "4.398000e-6158"
> in: /home/tuliom/src/libdfp/tests/test-bfp-conversions.c:333
> ```
> 
> We can see the first line is printing identical values for both the long double and the _Decimal128 (minus errors that did not appear after string conversions).
> But the expected and result outputs are completely different from what we expected.
> Those lines are printed by macro `_VC_P` which is expanded in line 610:
> 
> ```
>       _VC_P(__FILE__,tdtfp->line, tdtfp->e, retval, tdtfp->format);
> ```
> 
> Both `tdtfp->e` and `retval` are long double, with `tdtfp->e = 3.402819e+38L`.
> However `tdtfp->format = "%DDe"` and that's the format being used in `_VC_P` to print them.
> 
> ISO C11 states the following for `fprintf()`:
> 
> > If any argument is not the correct type for the corresponding conversion specification, the behavior is undefined.
> 
> In other words: test-bfp-conversions is broken and needs to be fixed and I'm working on a patch.

After reinstate the lines I've removed before those tests are passing. In fact I need to sum the second half of the number and making sense those tests are failing but, I guess what you're reporting here is another issue right? Maybe we have other failing results after fixing this :(

File: base-math/extendddtf.c [hunk 2]
```diff
@@ -30,20 +30,66 @@
...
 
-	df_part1 = a;			/* TD -> DF  */
-	dd_part1 = df_part1;		/* DF -> DD./TD.  */
-	dd_part2 = a - dd_part1;	/* DD./TD.  */
-	df_part2 = dd_part2;		/* DD/TD -> DF.  */
+     /* Get normalized exponent.  */
+    (void) getmantandexpd128 (a, &exp, 17, 1e17DL);
+
+    /* Check if the normalized exponent is beyond the exponent table and
+       round correct the result.  */
```
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: Yes. Is the maximum exponent represented by IBM long double (also is the same max exponent for double). Changed to "beyond maximum IBM long double exponent"

[the remark this answers is not in the document]
Comment (tuliom) [reply base-math/extendddtf.c]: @racardoso I'm afraid I was not clear in my comment: I think you should change to comment to make it clear.  ;-)

File: base-math/extendddtf.c [hunk 3]
```diff
@@ -30,20 +30,66 @@
...
+            case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -0.0 : 0.0;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -__LDBL_DENORM_MIN__ : 0.0;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -0.0 : __LDBL_DENORM_MIN__;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -0.0 : 0.0;
+          }
```
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: Fixed on my new commit.

File: base-math/extendddtf.c [hunk 4]
```diff
@@ -30,20 +30,66 @@
...
+            case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -LDBL_MAX : LDBL_MAX;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -INFINITY : LDBL_MAX;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -LDBL_MAX : INFINITY;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -INFINITY : INFINITY;
+          }
```
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: Fixed on my new commit.

File: base-math/extendddtf.c [hunk 5]
```diff
@@ -30,20 +30,66 @@
...
-	df_part2 = dd_part2;		/* DD/TD -> DF.  */
+     /* Get normalized exponent.  */
+    (void) getmantandexpd128 (a, &exp, 17, 1e17DL);
+
+    /* Check if the normalized exponent is beyond the exponent table and
+       round correct the result.  */
+    if (exp > BINPOWOF10_LIMIT)
+      {
+	    if (DFP_EXCEPTIONS_ENABLED)
+	      DFP_HANDLE_EXCEPTIONS (FE_OVERFLOW|FE_INEXACT);
```
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: Fixed on my new commit.

File: base-math/extendddtf.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: @tuliom That what I mean. I changed the comment to make it more clear

**Comment by racardoso** [comment]
@tuliom  @gftg85  bump. (force pushed)

File: base-math/extendddtf.c [hunk 6]
```diff
@@ -30,20 +30,68 @@
...
+	        case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -0.0 : 0.0;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -__LDBL_DENORM_MIN__ : 0.0;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -0.0 : __LDBL_DENORM_MIN__;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -0.0 : 0.0;
+	       }
```
Comment (tuliom) [inline base-math/extendddtf.c]: Likewise

File: base-math/extendddtf.c [hunk 7]
```diff
@@ -30,20 +30,68 @@
...
+	  }
+
+	df_part1 = a;			  /* TD -> DF  */
```
Comment (tuliom) [inline base-math/extendddtf.c]: This comment has an extra tab after `;`.

**Comment by racardoso** [comment]
@tuliom I guess now I fixed everything. Check out my new push

**Comment by tuliom** [comment]
> I guess what you're reporting here is another issue right?

@racardoso Exactly.  I'm fixing it with PR #80 .

File: base-math/extendddtf.c [hunk 8]
```diff
@@ -30,20 +30,68 @@
...
 CONVERT_WRAPPER(
 // extendddtf, trunctdtf
 	volatile double df_part1, df_part2;
+	int exp;
 	SRC_TYPE dd_part1, dd_part2;
 
-	df_part1 = a;			/* TD -> DF  */
-	dd_part1 = df_part1;		/* DF -> DD./TD.  */
-	dd_part2 = a - dd_part1;	/* DD./TD.  */
-	df_part2 = dd_part2;		/* DD/TD -> DF.  */
```
[the remark this answers is not in the document]
Comment (racardoso) [reply base-math/extendddtf.c]: @gftg85 ok. Force pushed again with your sugestions

File: base-math/extendddtf.c [hunk 9]
```diff
@@ -30,14 +30,61 @@
...
+#include "convert_helpers.h"
 
 CONVERT_WRAPPER(
 // extendddtf, trunctdtf
 	volatile double df_part1, df_part2;
+	int exp;
 	SRC_TYPE dd_part1, dd_part2;
 
+	/* Get normalized exponent.  */
+	(void) getmantandexpd128 (a, &exp, 17, 1e17DL);
```
Comment (gftg85) [inline base-math/extendddtf.c]: Is there a reason why your have to pass `shift = 17` and `_y = 1e17DL`, here?  In other calls to `getmantandexpd128`, such as in `trunctdsf.c` and `trunctddf.c`, the return of `getmantandexpd128` is actually used, so it makes sense to perform the shift.  Here, the shift has no effect, since you discard the return of `getmantandexpd128`.

For that reason, and in order to avoid a future misunderstanding, I think you should convert this line into:

```
(void) getmantandexpd128 (a, &exp, 0, 0);
```

Comment (racardoso) [reply base-math/extendddtf.c]: @gftg85  No special reason in fact I am working to not call this function at all since there's lot's of stuff that I don't need there. But, I will send that in another PR. For now I already changed to your sugestion

Comment (gftg85) [reply base-math/extendddtf.c]: Thanks.
